mlx-nn/src/optimizers/rmsprop.rs

```diff
@@ -0,0 +1,141 @@
+use std::rc::Rc;
+
+use mlx_rs::{
+    array,
+    ops::{sqrt, square},
+    Array,
+};
+
+use crate::utils::get_mut_or_insert_with;
+
+use super::*;
+
+/// The RMSprop optimizer [1].
+///
+/// [1]: Tieleman, T. and Hinton, G. 2012. Lecture 6.5-rmsprop, coursera: Neural networks for
+///     machine learning
+#[derive(Debug, Clone)]
+pub struct RmsProp {
+    /// Learning rate
+    pub lr: f32,
+
+    /// The smoothing constant. Default to [`RmsProp::DEFAULT_ALPHA`] if not specified.
+    pub alpha: f32,
+
+    /// The epsilon added to the denominator to improve numerical stability. Default to
+    /// [`RmsProp::DEFAULT_EPSILON`] if not specified.
+    pub epsilon: f32,
+
+    /// Inner state
+    pub state: OptimizerState,
+}
+
+impl RmsProp {
+    /// Default alpha if not specified.
+    pub const DEFAULT_ALPHA: f32 = 0.99;
+
+    /// Default epsilon if not specified.
+    pub const DEFAULT_EPSILON: f32 = 1e-8;
+
+    /// Creates a new `RmsProp` optimizer.
+    pub fn new(lr: f32) -> Self {
+        Self {
+            lr,
+            alpha: Self::DEFAULT_ALPHA,
+            epsilon: Self::DEFAULT_EPSILON,
+            state: OptimizerState::new(),
+        }
+    }
+
+    /// Sets the smoothing constant. Default to [`RmsProp::DEFAULT_ALPHA`] if not specified.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `alpha` is negative.
+    pub fn with_alpha(mut self, alpha: impl Into<Option<f32>>) -> Self {
```

Comment:
Should we type this to not be negative? I'm always a bit torn on that kind of typing

Reply:
Is there an existing crate that does this for `f32` or should we add a new type?